### ISSUE TITLE
[ROS2] Multiple definitions linking errors for functions defined in header files

### ISSUE THREAD

**Issue by Marwan99** [ros/filters#42]
Related to [ANYbotics/grid_map#266 ](https://github.com/ANYbotics/grid_map/pull/266)

I am trying to build a [library](https://github.com/ANYbotics/grid_map/blob/6e3305664854b27946ea0921b022e327a6c4d0d5/grid_map_filters/CMakeLists.txt#L49-L66) with multiple filter plugins, where in each filter [filter_base.hpp](https://github.com/ros/filters/blob/ros2/include/filters/filter_base.hpp) is being included. I am getting linking errors ([CI output](https://app.circleci.com/pipelines/github/ANYbotics/grid_map/171/workflows/05dff0bf-7625-4d92-94b6-9b015704a64e/jobs/178)) complaining about multiple definitions of [filters::impl::normalize_param_prefix](https://github.com/ros/filters/blob/bfb591760164ecf71542778fdd59833afcca8aa7/include/filters/filter_base.hpp#L47). Same thing is happening with [filters::impl::load_chain_config](https://github.com/ros/filters/blob/bfb591760164ecf71542778fdd59833afcca8aa7/include/filters/filter_chain.hpp#L62-L63).

I am getting this behaviour on ROS2 Rolling using bfb5917 and on Foxy using the binaries.

Exact errors:
```
/usr/bin/ld: CMakeFiles/grid_map_filters.dir/src/MinInRadiusFilter.cpp.o: in function `filters::impl::normalize_param_prefix(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)':
MinInRadiusFilter.cpp:(.text+0xc0): multiple definition of `filters::impl::normalize_param_prefix(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'; CMakeFiles/grid_map_filters.dir/src/ThresholdFilter.cpp.o:ThresholdFilter.cpp:(.text+0xc0): first defined here

```

```
/usr/bin/ld: CMakeFiles/filters_demo.dir/src/FiltersDemo.cpp.o: in function `filters::impl::load_chain_config(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<rclcpp::node_interfaces::NodeLoggingInterface> const&, std::shared_ptr<rclcpp::node_interfaces::NodeParametersInterface> const&, std::vector<filters::impl::FoundFilter, std::allocator<filters::impl::FoundFilter> >&)':
FiltersDemo.cpp:(.text+0xd00): multiple definition of `filters::impl::load_chain_config(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<rclcpp::node_interfaces::NodeLoggingInterface> const&, std::shared_ptr<rclcpp::node_interfaces::NodeParametersInterface> const&, std::vector<filters::impl::FoundFilter, std::allocator<filters::impl::FoundFilter> >&)'; CMakeFiles/filters_demo.dir/src/filters_demo_node.cpp.o:filters_demo_node.cpp:(.text+0x210): first defined here
```

I think this could be fixed be making the functions mentioned above **static** as to make their scope local to their receptive files, does that sound right?

Please let me know if you need any further information, thanks in advance for your help :)

**Comment by SteveMacenski** [comment]
@Marwan99 mark free functions as `inline` to handle this.

**Comment by Marwan99** [comment]
@SteveMacenski Thanks for the suggestion, can confirm that it solves the linking errors. Will submit a PR with this fix.

**Comment by tfoote** [comment]
Fixed in #43
